pkg/controller/populators/clone-populator.go

```diff
@@ -0,0 +1,379 @@
+	controller controller.Controller
+	planner    *clone.Planner
+}
+
+// NewClonePopulator creates a new instance of the clone-populator controller
+func NewClonePopulator(
+	ctx context.Context,
+	mgr manager.Manager,
+	log logr.Logger,
+	clonerImage string,
+	pullPolicy string,
+	installerLabels map[string]string,
+) (controller.Controller, error) {
+	client := mgr.GetClient()
+	reconciler := &ClonePopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(clonePopulatorName),
+			recorder:        mgr.GetEventRecorderFor(clonePopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			sourceKind:      cdiv1.VolumeCloneSourceRef,
+			installerLabels: installerLabels,
+		},
+		manager: mgr,
+	}
+
+	clonePopulator, err := controller.New(clonePopulatorName, mgr, controller.Options{
+		MaxConcurrentReconciles: 5,
+		Reconciler:              reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+	reconciler.controller = clonePopulator
+
+	reconciler.planner = &clone.Planner{
+		RootObjectType:  &corev1.PersistentVolumeClaimList{},
+		OwnershipLabel:  LabelOwnedByUID,
+		UIDField:        uidField,
+		Image:           clonerImage,
+		PullPolicy:      corev1.PullPolicy(pullPolicy),
+		InstallerLabels: installerLabels,
+		Client:          reconciler.client,
+		Recorder:        reconciler.recorder,
+		Controller:      reconciler.controller,
+	}
+
+	if err := addCommonPopulatorsWatches(mgr, clonePopulator, log, cdiv1.VolumeCloneSourceRef, &cdiv1.VolumeCloneSource{}); err != nil {
+		return nil, err
+	}
+
+	if err := reconciler.planner.AddCoreWatches(reconciler.log); err != nil {
+		return nil, err
+	}
+
+	return clonePopulator, nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of VolumeCloneSource kind
+func (r *ClonePopulatorReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Clone Source PVC")
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(ctx, req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// replace this eventually
+	if pvc.Spec.DataSourceRef != nil &&
+		pvc.Spec.DataSourceRef.Namespace != nil &&
+		*pvc.Spec.DataSourceRef.Namespace != pvc.Namespace {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, fmt.Errorf("cross namespace datasource not supported yet"))
+	}
+
+	hasFinalizer := cc.HasFinalizer(pvc, cloneFinalizer)
+	isBound := pvc.Spec.VolumeName != ""
+	isDeleted := !pvc.DeletionTimestamp.IsZero()
+
+	log.V(1).Info("pvc state", "hasFinalizer", hasFinalizer, "isBound", isBound, "isDeleted", isDeleted)
+
+	if !isDeleted && !isBound {
+		return r.reconcilePending(ctx, log, pvc)
+	}
+
+	if (isBound || isDeleted) && hasFinalizer {
```

Comment:
No, normally, the pvc will only have the finalizer until it is populated.  After that the pvc may live a long life with no finalizer.  Eventually it will get deleted and not have a finalizer

Reply:
got it